### PR TITLE
VL_USelect-TS-migration_Vitalii-Dudnik

### PR DESCRIPTION
**Pull request by KinduD21**
Migrate USelect component to TypeScript

### CODE REVIEW

File: src/ui.form-select/USelect.vue [hunk 1]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
 const preferredOpenDirection = ref(DIRECTION.bottom);
 const search = ref("");
 
-const dropdownListRef = ref(null);
-const wrapperRef = ref(null);
-const searchInputRef = ref(null);
-const labelComponentRef = ref(null);
-const leftSlotWrapperRef = ref(null);
-const innerWrapperRef = ref(null);
+const dropdownListRef = ref<DropdownListRef | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:90]: You don't need custom DropdownListRef type, detete it and use [ComponentExposed](https://github.com/vuelessjs/vueless/blob/main/src/types.ts#L477) utility type. Look for examples in other components.

File: src/ui.form-select/USelect.vue [hunk 2]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
...
+const wrapperRef = ref<HTMLElement | null>(null);
+const searchInputRef = ref<HTMLElement | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:92]: HTMLInputElement

File: src/ui.form-select/USelect.vue [hunk 3]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
...
 const search = ref("");
 
-const dropdownListRef = ref(null);
-const wrapperRef = ref(null);
-const searchInputRef = ref(null);
-const labelComponentRef = ref(null);
-const leftSlotWrapperRef = ref(null);
-const innerWrapperRef = ref(null);
+const dropdownListRef = ref<DropdownListRef | null>(null);
+const wrapperRef = ref<HTMLElement | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:91]: HTMLDivElement

File: src/ui.form-select/USelect.vue [hunk 4]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
...
+const searchInputRef = ref<HTMLElement | null>(null);
+const labelComponentRef = ref<{ labelElement: HTMLElement } | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:93]: Use [ComponentExposed](https://github.com/vuelessjs/vueless/blob/main/src/types.ts#L477)

File: src/ui.form-select/USelect.vue [hunk 5]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
...
+const leftSlotWrapperRef = ref<HTMLElement | null>(null);
+const innerWrapperRef = ref<HTMLElement | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:95]: HTMLDivElement

File: src/ui.form-select/USelect.vue [hunk 6]
```diff
@@ -585,12 +87,17 @@ const isOpen = ref(false);
...
-const wrapperRef = ref(null);
-const searchInputRef = ref(null);
-const labelComponentRef = ref(null);
-const leftSlotWrapperRef = ref(null);
-const innerWrapperRef = ref(null);
+const dropdownListRef = ref<DropdownListRef | null>(null);
+const wrapperRef = ref<HTMLElement | null>(null);
+const searchInputRef = ref<HTMLElement | null>(null);
+const labelComponentRef = ref<{ labelElement: HTMLElement } | null>(null);
+const leftSlotWrapperRef = ref<HTMLElement | null>(null);
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:94]: HTMLSpanElement

File: src/ui.form-select/USelect.vue [hunk 7]
```diff
@@ -599,34 +106,33 @@ const isTop = computed(() => {
...
 
 const dropdownValue = computed({
   get: () => props.modelValue,
   set: (newValue) => {
-    let value = newValue;
+    let value;
 
     if (props.multiple) {
-      value = Array.isArray(props.modelValue) ? [...props.modelValue, newValue] : [newValue];
+      const multipleValue = Array.isArray(props.modelValue) ? props.modelValue : [];
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:131]: Wasn't it much cleaner before? This line look useful.

File: src/ui.form-select/USelect.vue [hunk 8]
```diff
@@ -599,34 +106,33 @@ const isTop = computed(() => {
...
+
+      value = [...multipleValue, newValue].flat();
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:133]: Why flat needed?

File: src/ui.form-select/USelect.vue [hunk 9]
```diff
@@ -795,30 +312,38 @@ function adjustPosition() {
   }
 }
 
-function removeElement(option, shouldClose = true) {
+function onMouseDownClearItem(event: MouseEvent, option: Option) {
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:315]: Why clear? Looks like it removes item at all.

File: src/ui.form-select/USelect.vue [hunk 10]
```diff
@@ -775,13 +290,15 @@ function activate() {
     nextTick(() => searchInputRef.value && searchInputRef.value.focus());
   }
 
-  if (wrapperRef.value !== undefined && !props.searchable) wrapperRef.value.focus();
+  if (wrapperRef.value !== null && !props.searchable) {
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:293]: ` if (wrapperRef.value && !props.searchable) {`

File: src/ui.form-select/USelect.vue [hunk 11]
```diff
@@ -642,55 +148,76 @@ const isSelectedValueLabelVisible = computed(() => {
...
 
 const isLocalValue = computed(() => {
-  if (props.multiple) return Boolean(localValue.value.length);
+  const value = localValue.value;
 
-  return typeof localValue.value !== "number"
-    ? Boolean(localValue.value)
-    : Boolean(String(localValue.value));
+  if (Array.isArray(value)) {
+    return !!value?.length;
```
Comment (Explicit12) [inline src/ui.form-select/USelect.vue:203]: `Boolean(value.length);`

It looks like a bunch of symbols now. And you don't need optional chaining operator here, array always have length property.

**Review by Explicit12 — CHANGES_REQUESTED** [review]

File: src/ui.form-select/USelect.vue (hunk 7 above)
Comment (itsJohnnyGrid) [reply src/ui.form-select/USelect.vue:131]: we need to apply `flat()` and by this way it looks more readable.

You can suggest resulting piece of code in comments below.

File: src/ui.form-select/USelect.vue (hunk 11 above)
Comment (itsJohnnyGrid) [reply src/ui.form-select/USelect.vue:203]: !! is more readable and consistens in this case.

image

File: src/ui.form-select/USelect.vue (hunk 9 above)
Comment (Explicit12) [reply src/ui.form-select/USelect.vue:313]: I'd name it onMouseDownClearSelectedItem than.

**Review by Explicit12 — APPROVED** [review]

**Pull request merged**